test/Test/Pos/Ssc/GodTossing/Toss/BaseSpec.hs

```diff
@@ -0,0 +1,573 @@
+checksGoodSharesPayload :: GoodSharesPayload -> Bool
+checksGoodSharesPayload (getGoodShares -> (epoch, gtgs, sharesMap, mrs)) =
+    case tossRunner mrs gtgs $ checkSharesPayload epoch sharesMap of
+        Left (DecrSharesNotMatchCommitment _) -> True
+        Right _ -> True
+        _ -> False
+
+-- | Checks that when the data 'checkSharesPayload' is passed is incorrect w.r.t.
+-- 'checkSharesPayload', said function fails.
+-- NOTE: does not check for 'DecrSharesNotMatchCommitment' failure. This would make the
+-- already non-trivial arbitrary instance for 'GoodSharesPayload' unmanageable.
+checksBadSharesPayload
+    :: GoodSharesPayload
+    -> StakeholderId
+    -> NonEmpty (AsBinary Share)
+    -> VssCertificate
+    -> Property
+checksBadSharesPayload
+    (getGoodShares -> (epoch, g@GtGlobalState {..}, sm, mrs))
+    sid
+    ne
+    cert =
+    let -- This property assumes the existence of a stakeholder not in the commitments or
+        -- shares map. Instead of writing a new 'Arbitrary' type which will only be useful
+        -- here, this condition is just enforced by deleting it from where it shouldn't
+        -- be.
+        gtgs = g & (gsShares %~ HM.delete sid) .
+                   (gsCommitments %~ mkCommitmentsMapUnsafe .
+                                     HM.delete sid .
+                                     getCommitmentsMap)
+        sharesMap = fmap (HM.delete sid) . HM.delete sid $ sm
+
+        mrsWithMissingEpoch = HM.delete epoch mrs
+        noRichmen =
+            tossRunner mrsWithMissingEpoch gtgs $ checkSharesPayload epoch sharesMap
+        res1 = case noRichmen of
+            Left (NoRichmen e) -> e == epoch
+            _ -> False
+
+        newSharesMap = HM.insert sid mempty sharesMap
+        sharesNotRichmen = tossRunner mrs gtgs $ checkSharesPayload epoch newSharesMap
+        res2 = case sharesNotRichmen of
+            Left (SharesNotRichmen nes) -> sid `elem` nes
+            _ -> False
+
+        newerSharesMap = fmap (HM.insert sid ne) sharesMap
+        internalShareWithoutComm =
+            tossRunner mrs gtgs $ checkSharesPayload epoch newerSharesMap
+        res3 = case internalShareWithoutComm of
+            Left (InternalShareWithoutCommitment nes) -> sid `elem` nes
+            _ -> False
+
+        newestSharesMap = HM.insert sid mempty sharesMap
+        gtgs' = gtgs & (gsShares %~ HM.insert sid mempty) .
+                       (gsVssCertificates %~ \vcd@VssCertData{..} ->
+                           vcd { certs = HM.insert sid cert certs})
+        mrs' = HM.update (Just . HM.insert sid (mkCoin 0)) epoch mrs
+        sharesAlreadySent =
+            tossRunner mrs' gtgs' $ checkSharesPayload epoch newestSharesMap
+        res4 = case sharesAlreadySent of
+            Left (SharesAlreadySent nes) -> sid `elem` nes
+            _ -> False
+
+    in not (HM.member sid _gsShares) && not (null sharesMap) ==>
```

Comment:
Oh, good catch. Essentially, I do not enforce the absence of any particular stakeholder in the `Arbitrary` instances, so if I want these tests that check failure to work I have to delete the `StakeholderId` passed as an argument to the property. Otherwise, if the stakeholder is, by chance, in the global state, one of the failures won't be triggered and the property won't work.